libyul/optimiser/UnusedStoreEliminator.cpp

```diff
@@ -0,0 +1,446 @@
+				state = State::Used;
+			else if (_operation.effect == Effect::Write && knownCovered(storeOperation, _operation))
+				state = State::Unused;
+		}
+}
+
+bool UnusedStoreEliminator::knownUnrelated(
+	UnusedStoreEliminator::Operation const& _op1,
+	UnusedStoreEliminator::Operation const& _op2
+) const
+{
+	KnowledgeBase knowledge(m_dialect, m_ssaValues);
+
+	if (_op1.location != _op2.location)
+		return true;
+	if (_op1.location == Location::Storage)
+	{
+		if (_op1.start && _op2.start)
+			return knowledge.knownToBeDifferent(*_op1.start, *_op2.start);
+	}
+	else
+	{
+		u256 largestPositive = (u256(1) << 128) - 1;
+
+		yulAssert(_op1.location == Location::Memory, "");
+		if (_op1.length && knowledge.knownToBeZero(*_op1.length))
+			return true;
+		if (_op2.length && knowledge.knownToBeZero(*_op2.length))
+			return true;
+
+		// 1.start + 1.length <= 2.start
+		if (_op1.length && _op2.start && _op1.start)
+		{
+			optional<u256> length1 = knowledge.valueIfKnownConstant(*_op1.length);
+			optional<u256> diff = knowledge.differenceIfKnownConstant(*_op2.start, *_op1.start);
+			// diff = 2.start - 1.start
+			if (length1 && diff)
+				if (
+					*length1 <= *diff &&
+					*length1 <= largestPositive &&
+					*diff <= largestPositive
+				)
+					return true;
+		}
+		// 2.start + 2.length <= 1.start
+		if (_op2.length && _op1.start && _op2.start)
+		{
+			optional<u256> length2 = knowledge.valueIfKnownConstant(*_op2.length);
+			optional<u256> diff = knowledge.differenceIfKnownConstant(*_op1.start, *_op2.start);
+			// diff = 1.start - 2.start
+			if (length2 && diff)
+				if (*length2 <= *diff && *length2 <= largestPositive && *diff <= largestPositive)
+					return true;
+		}
+	}
+
+	return false;
+}
+
+bool UnusedStoreEliminator::knownCovered(
+	UnusedStoreEliminator::Operation const& _covered,
+	UnusedStoreEliminator::Operation const& _covering
+) const
+{
+	if (_covered.location != _covering.location)
+		return false;
+	if (
+		(_covered.start && _covered.start == _covering.start) &&
+		// length is unused for storage
+		(_covered.location == Location::Storage || (_covered.length && _covered.length == _covering.length))
+	)
+		return true;
+	if (_covered.location == Location::Memory)
+	{
+		KnowledgeBase knowledge(m_dialect, m_ssaValues);
+		u256 largestPositive = (u256(1) << 128) - 1;
+
+		if (_covered.length && knowledge.knownToBeZero(*_covered.length))
+			return true;
+		// Condition (i = cover_i_ng, e = cover_e_d):
+		// i.start <= e.start && e.start + e.length <= i.start + i.length
+		if (!_covered.start || !_covering.start || !_covered.length || !_covering.length)
+			return false;
+		optional<u256> startDiff = knowledge.differenceIfKnownConstant(*_covered.start, *_covering.start);
+		optional<u256> coveredLength = knowledge.valueIfKnownConstant(*_covered.length);
+		optional<u256> lengthDiff = knowledge.differenceIfKnownConstant(*_covering.length, *_covered.length);
+		if (
+			(startDiff && coveredLength && lengthDiff) &&
```

Comment:
We should cehck this, too.

Reply:
What about that Z3 proof you mentioned earlier on the chat?